scripts/RPS-optimized.js

```diff
@@ -1,33 +1,20 @@
 /*
-    Note: this is not a real-life example unless you want to test for the absolute RPS limit of your system.
-    This is extremely aggressive test.
+  Note: this is not a real-life example unless you want to test for the absolute RPS limit of your system.
+  This is extremely aggressive test.
 */
 
 import http from "k6/http";
-import { sleep } from "k6";
 
 export let options = {
-	discardResponseBodies: true,
-    ext: {
-        loadimpact: {
-            name: `${__ENV.TEST_NAME}` || "RPS hammering test",
-            projectID: 3478725,
-        }
+  discardResponseBodies: true,
+  ext: {
+    loadimpact: {
+      name: `${__ENV.TEST_NAME}` || "RPS hammering test",
+      projectID: 3478725,
     }
+  }
 };
 
 export default function() {
-    // relentless hammering of these 6 URLs.
-    // 20 batches per iteration.
-    for(let i=0; i<20; i++){
-        http.batch([
-            ['GET', "http://34.249.156.120/static/logo.svg?ip=1"],
-            ['GET', "http://34.250.250.204/static/logo.svg?ip=2"],
-            ['GET', "http://52.31.47.157/static/logo.svg?ip=3"],
-            ['GET', "http://54.72.16.236/static/logo.svg?ip=4"],
-            ['GET', "http://54.76.206.198/static/logo.svg?ip=5"],
-            ['GET', "http://99.81.83.131/static/logo.svg?ip=6"],
-        ]);
-    }
-  sleep(0.1); // 100ms sleep between iterations.
+  http.get('http://test.staging.k6.io/static/logo.svg');
```

Comment:
Out of the curriocity, why we switching from the set of batches to a single request per iteration without sleep? :thinking:

Reply:
Good question :slightly_smiling_face: I should've made that clearer in the [commit message](https://github.com/grafana/k6-hardware-benchmark/commit/625566a2da6fa239d3f37cd239ddbe156753334d).

But basically, we don't have a set of IPs to make requests to as before, and `test.staging.k6.io` always resolves to 1 IP. So if we use the same hostname in an `http.batch()` call, we would quickly exhaust the ephemeral port range for that single IP. Remember that there's a 5-element tuple for establishing a network socket, and since the source port is the only variable in that case, we quickly reach the 65,536 limit (actually, a bit less than that), and get `cannot assign requested address` errors.

One way to avoid this is to make requests from different local IPs, using the [`--local-ips`](https://k6.io/docs/using-k6/k6-options/reference/#local-ips) option. Unfortunately, I wasn't able to establish connections from EC2 after adding more IPs, probably due to some missing configuration or routing (I was able to make it work on my local machine, though, so not sure what I was missing on EC2). In any case, doing this would complicate the amount of local setup we have to do before this test, and since we want to automate this from GitHub Actions, it's likely we wouldn't be able to make it work there anyway.

The drawback of this is that with `http.batch()` k6 can likely make more than the current ~161K RPS with just `http.get()`. Pawel was able to reach 188K with v0.26.2, and that was with a 0.1s sleep per iteration.

I think the solution should be to make the SUT available on several IPs, by either `test.staging.k6.io` returning more than 1 IP, or letting us know the specific IPs we should use, if we want to bypass the load balancer, like we did before. Having a set of fixed IPs is not maintainable though, so resolving to more than 1 IP would be the way to go. Unfortunately, I didn't get much help with this from Slava, and they have their own priorities to focus on, so I didn't keep pushing back.

For the purposes of these benchmarks, though, the point is not so much to reach the maximum possible number, but to set a performance baseline that we can use to compare with future results. There will undoubtedly be some variability, depending on the SUT response time, etc., but we should be able to get reliable results to determine if there are any performance changes in k6.

Hopefully that explains it, and let me know if you have any suggestions.

Reply:
Thanks for the great explanation! :bow: 

>Pawel was able to reach 188K with v0.26.2, and that was with half as many VUs (30K vs 60K), and a 0.1s sleep per iteration. 

Iteresting, what are the results for the `v0.26.2` on the same (current version) of test script.

>For the purposes of these benchmarks, though, the point is not so much to reach the maximum possible number, but to set a performance baseline that we can use to compare with future results. 

Yeah, I agree with you that we need the baseline first of all and do the tests regularly. The small concern that I have in my mind is that if the bottleneck currently the SUT, that means that we can potentially catch only degradations, but not the improvements. But as you said we could adress this in the future with tweaking the SUT  as you described it in preveous paragraph.

Reply:
Good point, but the SUT doesn't seem to be the bottleneck in this case. It's likely that if we were able to make requests from multiple IPs, we could achieve higher RPS.

The thing is that we _shouldn't_ do any tweaks to the SUT in the future, otherwise it would invalidate some of these results. Ideally, the SUT performance and deployment would remain static, so that the only variation comes from different versions of k6. In practice, this is difficult to achieve, since latencies and response times will fluctuate, even against the same SUT and with the same hardware, which will affect the results. But hopefully the variability is within a certain range that we find useful to determine whether the change comes from k6 or not.

Reply:
I do think it might just be better to more or less rerun for some set of versions every time. 

This removes the case where the SUT changed or we deployed somethign else slightly differently. 

It might be enough to run the previous version to see it gets roughly the same result and if it does - decide the SUT is the same. And not have to rerun for all versions :thinking: